rules/policy.go

```diff
@@ -59,9 +59,202 @@ func (r *DefaultRuleRenderer) ProtoRulesToIptablesRules(protoRules []*proto.Rule
+			markBitsToSetInitially = markAllBlocksPass
+		} else {
+			// There are at least some positive matches.  Leave the AllBlocks bit
+			// set to 0.  The first positive match block will set it directly if it
+			// matches.
+			markBitsToSetInitially = 0
+		}
+		rs = append(rs,
+			iptables.Rule{
+				Action: iptables.SetMaskedMarkAction{
+					Mark: markBitsToSetInitially,
+					Mask: markAllBlocksPass | markThisBlockPass,
+				},
+			},
+		)
+		usingCIDRBlocks = true
+		doneFirstPositiveMatchBlock := false
+
+		appendCIDRMatchBlock := func(cidrs []string, srcOrDst srcOrDst) {
+			// Implementation a positive match requires us to implement a logical
+			// "or" operation within the block and then "and" that with the result from
+			// the previous block.
+			//
+			// As an optimization, if rendering the first block, we simply set the
+			// "AllBlocks" bit if one of our rules matches.
+			//
+			// If we're not the first block, that doesn't work since the "AllBlocks"
+			// bit may already be set.  In that case, we write to a scratch "ThisBlock"
+			// bit, calculate the "and" at the end of the block and write that back
+			// to the "AllBlocks" bit.
+			var markToSet uint32
+			if doneFirstPositiveMatchBlock {
+				// This isn't the first block, we need to use a scratch bit to
+				// store the result.
+				markToSet = markThisBlockPass
+			} else {
+				// Optimization: since we're the first block, directly use the
+				// "AllBlocks" bit to store our result.
+				markToSet = markAllBlocksPass
+			}
+
+			// Render the per-CIDR rules.
+			for _, cidr := range cidrs {
+				rs = append(rs, iptables.Rule{
+					Match:  srcOrDst.MatchNet(cidr),
+					Action: iptables.SetMarkAction{Mark: markToSet},
+				})
+			}
+			if doneFirstPositiveMatchBlock {
+				// This isn't the first block, copy the scratch bit back to the
+				// AllBlocks bit.
+				rs = append(rs, iptables.Rule{
+					Match:  iptables.Match().MarkClear(markThisBlockPass),
+					Action: iptables.ClearMarkAction{Mark: markAllBlocksPass},
+				})
+			}
+			doneFirstPositiveMatchBlock = true
+		}
+
+		appendNegatedCIDRMatchBlock := func(cidrs []string, srcOrDst srcOrDst) {
+			// To implement a negated match we emit a rule per CIDR that does a positive
+			// match on the CIDR and *clears* the AllMatches bit if the packet matches.
+			// This gives the desired "not any" behaviour.
+			for _, cidr := range cidrs {
+				rule := iptables.Rule{
+					Match:  srcOrDst.MatchNet(cidr),
+					Action: iptables.ClearMarkAction{Mark: markAllBlocksPass},
+				}
+				rs = append(rs, rule)
+			}
+		}
+
+		if len(ruleCopy.SrcNet) > 1 {
+			// More than one positive match on source IP, need to render a block.  We
+			// prioritise the positive match over the negative match and avoid
+			// rendering a block if it'd only contain one IP.
+			appendCIDRMatchBlock(ruleCopy.SrcNet, src)
+			// Since we're using a block for this, nil out the match.
+			ruleCopy.SrcNet = nil
+		}
```

Comment:
As discussed, it's not very golang-ish to distinguish nil and 0-length for slices since `for .. range` and `len()` don't distinguish them either.